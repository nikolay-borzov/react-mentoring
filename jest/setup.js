import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
