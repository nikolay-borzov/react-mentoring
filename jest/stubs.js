export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'http://via.placeholder.com/500x750',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre']
}

export const films = [film]
