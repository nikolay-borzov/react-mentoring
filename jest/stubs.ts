import { Film } from '../src/entities/film'

export const film: Film = {
  id: 1,
  title: 'Film title',
  tagline: 'Put that cookie down',
  vote_average: 7.5,
  vote_count: 667,
  release_date: '2018-05-04',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Once upon a time...',
  budget: 15000000,
  revenue: 59735548,
  genres: ['some genre', 'another genre']
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id: number) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
