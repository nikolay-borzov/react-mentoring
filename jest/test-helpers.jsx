import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
