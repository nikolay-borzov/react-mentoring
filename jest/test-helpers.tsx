import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
