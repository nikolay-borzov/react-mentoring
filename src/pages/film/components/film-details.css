@import '../../../styles/variables.css';

.film-details__image {
  display: flex;
  justify-items: center;
  flex: 1 1 auto;
  min-width: 300px;
  max-width: 500px;
}

.film-details__description {
  flex: 1 1 100%;
  padding-left: 2rem;
}

@media (--viewport-bigger) {
  .film-details__description {
    padding-left: 4rem;
  }
}

.film-details__rating {
  display: flex;
  justify-content: center;
  align-items: center;
  color: var(--color-primary);
  font-size: 120%;
  font-weight: bold;
  width: 2rem;
  height: 2rem;
  padding: 1.2rem;
  border-radius: 100%;
  border: solid 3px var(--color-primary);
  margin-left: 2rem;
}

@media (max-width: 600px) {
  .film-details {
    flex-direction: column;
  }

  .film-details__description {
    padding-left: 0;
  }
}

