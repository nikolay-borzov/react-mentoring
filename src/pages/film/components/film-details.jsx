import React from 'react'
import PropTypes from 'prop-types'

import './film-details.css'

import { ContentImage } from '../../../components'

FilmDetails.propTypes = {
  film: PropTypes.object.isRequired
}

export function FilmDetails(props) {
  return (
    <div className="film-details flex padding-content">
      <div className="film-details__image">
        <ContentImage src={props.film.poster_path} alt={props.film.title} />
      </div>

      <div className="film-details__description">
        <div className="flex flex-align-center">
          <h1>{props.film.title} </h1>
          <span className="film-details__rating">
            {props.film.vote_average}
          </span>
        </div>

        <p className="font-big">{props.film.genres.join(', ')}</p>

        <p className="font-bold">{props.film.release_date.substring(0, 4)}</p>

        <p className="font-big">{props.film.overview}</p>
      </div>
    </div>
  )
}
