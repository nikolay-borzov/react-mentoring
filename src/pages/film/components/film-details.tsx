import React from 'react'
import styled from 'styled-components'

import { ContentImage } from '../../../components'
import { media } from '../../../styles/media'
import { Film } from '../../../entities/film'

export interface FilmDetailsProps {
  film?: Film
}

const FilmDetailsStyled = styled.div`
  @media (max-width: 600px) {
    flex-direction: column;
  }
`

const FilmDetailsImage = styled.div`
  display: flex;
  justify-items: center;
  flex: 1 1 auto;
  min-width: 300px;
  max-width: 500px;
`

const FilmDetailsDescription = styled.div`
  flex: 1 1 100%;
  padding-left: 2rem;

  @media (max-width: 600px) {
    padding-left: 0;
  }

  ${media.bigger`padding-left: 4rem;`};
`

const FilmDetailsRating = styled.span`
  display: flex;
  justify-content: center;
  align-items: center;
  color: var(--color-primary);
  font-size: 120%;
  font-weight: bold;
  width: 2rem;
  height: 2rem;
  padding: 1.2rem;
  border-radius: 100%;
  border: solid 3px var(--color-primary);
  margin-left: 2rem;
`

export const FilmDetails: React.FunctionComponent<FilmDetailsProps> = props => {
  let content

  if (props.film && props.film.id) {
    const {
      poster_path: posterPath,
      title,
      vote_average: voteAverage,
      genres,
      release_date: releaseDate,
      overview
    } = props.film

    content = (
      <>
        <FilmDetailsImage>
          <ContentImage src={posterPath} alt={title} />
        </FilmDetailsImage>

        <FilmDetailsDescription>
          <div className="flex flex-align-center">
            <h1>{title}</h1>
            <FilmDetailsRating>{voteAverage}</FilmDetailsRating>
          </div>

          <p className="font-size-big">{genres.join(', ')}</p>

          <p className="font-bold">{releaseDate.substring(0, 4)}</p>

          <p className="font-size-big">{overview}</p>
        </FilmDetailsDescription>
      </>
    )
  } else {
    content = (
      <div className="error-message centered">Unable to load the movie</div>
    )
  }

  return (
    <FilmDetailsStyled className="flex padding-content">
      {content}
    </FilmDetailsStyled>
  )
}
