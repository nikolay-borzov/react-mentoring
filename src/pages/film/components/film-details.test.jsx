import React from 'react'

import { film } from '../../../../jest/stubs'
import { itRendersCorrectlyShallow } from '../../../../jest/test-helpers'

import { FilmDetails } from './film-details'

describe('FilmDetails component', () => {
  const props = {
    film: film
  }

  beforeEach(() => {
    props.film = film
  })

  describe('renders correctly', () => {
    itRendersCorrectlyShallow(
      () => <FilmDetails {...props} />,
      'when film is provided'
    )

    itRendersCorrectlyShallow(() => {
      props.film = null
      return <FilmDetails {...props} />
    }, `when film isn't provided`)
  })
})
