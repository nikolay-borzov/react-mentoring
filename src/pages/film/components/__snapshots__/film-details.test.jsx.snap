// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmDetails component renders correctly when film is provided 1`] = `
<FilmDetailsStyled
  className="flex padding-content"
>
  <React.Fragment>
    <FilmDetailsImage>
      <ContentImage
        alt="Film title"
        src="https://picsum.photos/g/300/450/?random"
        title=""
      />
    </FilmDetailsImage>
    <FilmDetailsDescription>
      <div
        className="flex flex-align-center"
      >
        <h1>
          Film title
        </h1>
        <FilmDetailsRating>
          7.5
        </FilmDetailsRating>
      </div>
      <p
        className="font-size-big"
      >
        some genre, another genre
      </p>
      <p
        className="font-bold"
      >
        2018
      </p>
      <p
        className="font-size-big"
      >
        Film overview
      </p>
    </FilmDetailsDescription>
  </React.Fragment>
</FilmDetailsStyled>
`;

exports[`FilmDetails component renders correctly when film isn't provided 1`] = `
<FilmDetailsStyled
  className="flex padding-content"
>
  <div
    className="error-message centered"
  >
    Unable to load the movie
  </div>
</FilmDetailsStyled>
`;
